knative-operator/pkg/controller/knativekafka/knativekafka_controller.go

```diff
@@ -280,29 +320,41 @@ func (r *ReconcileKnativeKafka) delete(instance *operatorv1alpha1.KnativeKafka)
 }
 
 func (r *ReconcileKnativeKafka) deleteKnativeKafka(instance *operatorv1alpha1.KnativeKafka) error {
-	manifest, err := r.buildManifest(instance)
+	manifest, err := r.buildManifest(instance, ManifestBuildAll)
 	if err != nil {
 		return fmt.Errorf("failed to build manifest: %w", err)
 	}
 
-	if err := r.transform(manifest, instance); err != nil {
-		return fmt.Errorf("failed to transform manifest: %w", err)
+	stages := []func(*mf.Manifest, *operatorv1alpha1.KnativeKafka) error{
+		r.transform,
+		r.deleteResources,
 	}
 
-	if err := manifest.Delete(); err != nil {
-		return fmt.Errorf("failed to delete KnativeKafka manifest: %w", err)
+	// Execute each stage in sequence until one returns an error
+	for _, stage := range stages {
+		if err := stage(manifest, instance); err != nil {
+			return err
+		}
 	}
 	return nil
 }
 
-func (r *ReconcileKnativeKafka) buildManifest(instance *operatorv1alpha1.KnativeKafka) (*mf.Manifest, error) {
+type manifestBuild int
+
+const (
+	ManifestBuildEnabledOnly manifestBuild = iota
+	ManifestBuildDisabledOnly
+	ManifestBuildAll
```

Comment:
No need to export these.

Reply:
Not a huge fan of having such a switch on a function btw, that's usually a big old cold-smell (just like boolean switches are).

I can't think of an easy way to not do that without introducing quite a bit of boilerplate, so maybe that's just fine.

Reply:
yeah, I don't like this too. I feel you. 

> I can't think of an easy way to not do that without introducing quite a bit of boilerplate, so maybe that's just fine.

Same here